wallet-new/server/Cardano/Wallet/API/V1/LegacyHandlers/Addresses.hs

```diff
@@ -35,22 +42,40 @@ handlers =  listAddresses
        :<|> newAddress
        :<|> getAddress
 
+-- | This is quite slow. What happens when we have 50k addresses?
+-- TODO(ks): One idea I have is to persist the length of the
+-- addresses and send that to `respondWith`,
+-- while we could lazily fetch the data (and have better performance)
+-- we need to show what the current page contains?
+-- In other words, we wouldn't send everything to `respondWith`.
+-- Another idea is to use actual paging to reduce the footprint of all
+-- these calls. The flaw with this is that we need to think about deletion,
+-- but I have an idea or two how that can be fixed.
 listAddresses
-    :: MonadIO m
-    => RequestParams -> m (WalletResponse [WalletAddress])
-listAddresses RequestParams {..} = do
-    -- TODO(matt.parsons):  Fix as part of CSL-2146
-    addresses <- liftIO $ generate (vectorOf 2 arbitrary)
-    return WalletResponse {
-              wrData = addresses
-            , wrStatus = SuccessStatus
-            , wrMeta = Metadata $ PaginationMetadata {
-                        metaTotalPages = 1
-                      , metaPage = 1
-                      , metaPerPage = 20
-                      , metaTotalEntries = 2
-                      }
-            }
+    :: forall ctx m. (MonadThrow m, V0.MonadWalletLogic ctx m)
+    => RequestParams
+    -> m (WalletResponse [WalletAddress])
+listAddresses params = do
+
+    wdb <- askWalletDB
+    ws  <- getWalletSnapshot wdb
+
+    let allAddresses = runStreamAddresses ws
+
+    respondWith params (NoFilters :: FilterOperations WalletAddress)
+                       (NoSorts   :: SortOperations   WalletAddress)
+                       allAddresses
+  where
+    -- | Should improve performance, stream fusion ultra super nuclear
+    -- fission... Insert cool word of coice.
+    runStreamAddresses :: WalletSnapshot -> m (IxSet' WalletAddress)
+    runStreamAddresses ws =
+        runConduit   $ CL.sourceList (getWalletAddresses ws)
+                    .| CL.map Just
+                    .| CL.concatMapM V0.getAccounts
+                    .| CL.concatMap caAddresses
+                    .| CL.mapM migrate
+                    .| CL.fold (\x a -> IxSet.updateIx (addrId a) a x) IxSet.empty
```

Comment:
The previous implementation was collecting the list into memory and then converting to `IxSet`. This implementation streams the elements directly into the `IxSet`.

It does assume that the `addrId` property is unique. But, shouldn't it be?

Reply:
yup, an `addrId` should be unique (I assume that's nothing less than a Cardano's `Address`), and therefore, globally unique (as in worldwide-unique 😁 )